Fitting and interpreting linear models (high volume)
====================================================

```{r include = FALSE}
library(knitr)
opts_chunk$set(tidy = FALSE)
```


```{r}
library(limma)
library(ggplot2) 
library(lattice)
library(reshape) # for the function melt
library(plyr)
prDat <- read.table("../data/photoRec/GSE4051_data.tsv")
str(prDat, max.level = 0)
prDes <- readRDS("../data/photoRec/GSE4051_design.rds")
str(prDes)
library(VennDiagram)
```

## Write a function to prepare a mini-dataset for a small number of genes

```{r}
prepareData <- function(genes, finishing=NULL){
  jDat <- prDat[genes, prDes$sidChar]
  newDat <- cbind(prDes, t(jDat))
  
  # in case of multiple probes selected
  newDat <- melt(newDat,
                 measure.vars=grep("_at", colnames(newDat), value=TRUE))
  newDat <- rename(newDat, c("value"="gExp",
                             "variable"="gene"))  # rename column
  
  # some finishing touch to the data before returning
  if(!is.null(finishing)){
    newDat <- finishing(newDat)
  }
  return(newDat)
}
```

```{r include=FALSE}
(luckyGenes <- c("1419655_at","1438815_at"))
jDat <- prepareData(luckyGenes)
str(jDat)
head(jDat)
tail(jDat)
```


```{r}
# a function to stripplot a mini-dataset
makeStripplot <- function(myDat, ...) {
  p <- ggplot(myDat, 
              aes(x=devStage, y=gExp, color=gType, group=gType))
  p <- p + facet_wrap(~gene) + geom_point() 
  p <- p + stat_summary(fun.y=mean, geom="line")
  # p <- p + geom_line(stat = "summary", fun.y = mean)  # alternatively
  return(p)
}
```


## The difficulty in estimating gene-wise variance
  
Before we dive in and start using `limma` with the `photoRec` dataset, let's do a small simulation to illustrate how lousy variance estimates can be when the number of samples is small.

Let's simulate data for 1000 genes. For each gene, we get 3 observations from a normal distribution with mean 0 and variance 1. We generate the data for each gene independent of the others.

```{r}
m <- 1000
n <- 3
x <- matrix(rnorm(m * n), nrow = m)
```

Let's take the observed gene-wise variances. Yes, folks, we are estimating variance with samples of size 3. People do this all the time -- remember the video? We inspect them a bit numerically and graphically.

```{r}
obsVars <- apply(x, 1, var)
summary(obsVars)
densityplot(~ obsVars, n = 200)
```

Notice how many of the observed variances are freakishly small (and freakishly large!), even though they are indeed equal to 1 "on average". For example, we see that at least a quarter of the genes appear to exhibit a sample variance that is less than one-third the true variance. This can wreak havoc with statistical inference, such as t-statistics. This is what `limma` -- or the statistical methods it embodies, actually -- is designed to combat.

Optional take-home exercise: Make the above simulation more realistic with two (or more) groups, different data-generating means and group differences, different data-generating gene-wise variances, etc.

```{r}
m <- 1000  # simulate 1000 genes
nGroup <- 4
groupSize <- m / nGroup
nSample <- 3
(groupMeans <- rnorm(nGroup, 0, 5))
x2 <- ldply(groupMeans, function(mu){
  matrix(rnorm(groupSize * nSample, mean=mu), 
         nrow = groupSize)
})
```

```{r}
obsVars2 <- apply(x2, 1, var)
summary(obsVars2)
densityplot(~ obsVars2, n = 200)
```


## Fit a linear model: explain gene expression in the wild type mice as a function of developmental stage (one-way ANOVA)

Let's just work with the wild type data.

```{r}
wtDes <- subset(prDes, gType == "wt")
wtDat <- subset(prDat, select = prDes$gType == "wt")
str(wtDat, max.level = 0)
```

```{r}
(wtDes <- droplevels(wtDes))  # not necessary here, but better get into habit of doing it
(wtDesMat <- model.matrix(~ devStage, wtDes))
```

```{r}
wtFit <- lmFit(wtDat, wtDesMat)
wtEbFit <- eBayes(wtFit)
```

The first thing we might ask is "which genes show differential expression over the course of development"? 

```{r}
dsHits <- topTable(wtEbFit, coef=grep("devStage", colnames(wtDesMat)),
         p.value=1e-5, number=Inf)
```


```{r}
makeStripplot(prepareData(rownames(dsHits)[c(3, 6, 9)],
                          finishing=function(x){subset(x, gType=="wt")}))
```

Consider the effects associated with developmental stages P2 and P10. Scatterplot the t statistics for the test that the P2 effect is zero against that for P10.
```{r}
P2tt <- topTable(wtEbFit, coef = "devStageP2", n = Inf, sort = "none")
P10tt <- topTable(wtEbFit, coef = "devStageP10", n = Inf, sort = "none")
# names(P2tt) <- paste("P2", names(P2tt), sep=".")
# names(P10tt) <- paste("P10", names(P10tt), sep=".")

splom(cbind(P2.t=P2tt$t, P10.t=P10tt$t), panel=panel.smoothScatter)
```

Create a densityplot of the associated adjusted p-values, so you can get a sense of which developmental stage, P2 or P10, is more clearly distinguished from baseline E16.

```{r}
tt2and10 <- rbind(cbind(P2tt, effect="P2", probe=rownames(P2tt)),
                  cbind(P10tt, effect="P10", probe=rownames(P10tt)))
ggplot(tt2and10, aes(x=adj.P.Val, group=effect, color=effect)) + 
  geom_density()
```

Is this what you'd expect? Yes, P10 is further away from E16 devStage used as intercept.

If you require a BH adjusted p-value less than 1e-03, how many hits do you get for P2? How many for P10? How much overlap is there?

```{r}
hits2and10 <- subset(tt2and10, adj.P.Val<1e-3)
hitGenes <- dlply(hits2and10, ~effect, function(x){x$probe})

# Start a new plot
plot.new()

# Draw the Venn diagram. Note the argument `filename=NULL` tells it to
# create a plot object instead of outputting to file.
venn.plot <- venn.diagram(hitGenes, filename = NULL, 
                          fill = c("blue", "yellow"))

# Draw the plot on the screen.
grid.draw(venn.plot)
```

Now just focus on the P10 effect. Create a scatterplot matrix of raw p-values, BH adjusted p-values, and BY p-values.

```{r}
P10pVals <- data.frame(raw = P10tt$P.Value,
                       BH = P10tt$adj.P.Val,
                       BY = p.adjust(P10tt$P.Value, method = "BY"))
splom(P10pVals,
      panel = function(x, y, ... ) {
          panel.xyplot(x, y, pch = ".", ...)
          panel.smoothScatter(x,y)
          panel.abline(a = 0, b = 1, col = "orange")
      })
```


## Perform inference for some contrasts

Let's try to distinguish genes that have stable expression at the last three developmental stages (P6, P10, and 4_weeks) from those that do not. If expression doesn't change from P6 to P10 to 4_weeks, then the effects for all 3 of those developmental stages should be the same. That means that the difference between the P10 and P6 effects is zero and ditto for the difference between 4_weeks effect and P10 (or P6, for that matter). Let's form these contrasts.
```{r}
colnames(wtDesMat)
(cont.matrix <- makeContrasts(
    P10VsP6 = devStageP10 - devStageP6,
    fourweeksVsP10 = devStage4_weeks - devStageP10,
    levels = wtDesMat))
wtFitCont <- contrasts.fit(wtFit, cont.matrix)
wtEbFitCont <- eBayes(wtFitCont)
```
What does `topTable()` do with our contrasts?

Something like using the previous fitted model to get new parameters and statistics based on some transformations.

The top hits are probes where there is big change from P6 to P10, from P10 to 4_weeks, or both. Let's check that by plotting the data from the top 4 hits.

```{r}
hits0 <- topTable(wtEbFitCont)
makeStripplot(prepareData(rownames(hits0)[1:4]))
```

So far, so good. These 4 probes show little expression change from P6 to P10 and a strong increase from P10 to 4_weeks. I would like to find some where there's a change in each case but perhaps in opposite direction. Let's press on.

Let's use `decideTests()` to adjust the p-values for both contrasts globally, i.e. all together and then threshhold them at a cutoff of 1e-04.
```{r}
cutoff <- 1e-04
wtResCont <- decideTests(wtEbFitCont, p.value = cutoff, method = "global")
summary(wtResCont)
```
We see there are 4 probes that go down from P6 to P10 and no hits going the other way. There are 8 probes that go down from P10 to 4_weeks and 46 going the other way. Let's try to pull out various hits and plot their data.

Here are the 4 that decline from P6 to P10.
```{r}
(hits1 <- rownames(prDat)[wtResCont[, "P10VsP6"] < 0])
makeStripplot(prepareData(hits1))
```

Here are 4 of the 8 that decline from P10 to 4_weeks.
```{r}
(hits2 <- rownames(prDat)[wtResCont[, "fourweeksVsP10"] < 0])
makeStripplot(prepareData(hits2[1:4]))
```

Is there any overlap between these probes?
```{r}
intersect(hits1, hits2)
```
Apparently not.

Here are 4 of the 46 that increase from P10 to 4_weeks.

```{r}
(hits3 <- rownames(prDat)[wtResCont[, "fourweeksVsP10"] > 0])
stripplotIt(prepareData(hits3[1:4]))
```

Is there any overlap between these probes and the previous "down" hits?

```{r}
intersect(hits1, hits3)
intersect(hits2, hits3)
```


## Take-home exercise

See if you can find one or more probes that have some expression changes up to P6 and then hold steady all the way to 4_weeks. 
```{r}
# identify changes up to P6
colnames(wtDesMat)
(cont.matrix2 <- makeContrasts(
    P2VsE16 = devStageP2,
    P6VsP2 = devStageP6 - devStageP2,
    P6VsE16 = devStageP6,
    levels = wtDesMat))
wtFitCont2 <- contrasts.fit(wtFit, cont.matrix2)
wtEbFitCont2 <- eBayes(wtFitCont2)

hits20 <- topTable(wtEbFitCont2)
makeStripplot(prepareData(rownames(hits20)[1:4]))
```

Let's use `decideTests()` to adjust the p-values for both contrasts globally, i.e. all together and then threshhold them
```{r}
cutoff <- 1e-04
wtResCont2 <- decideTests(wtEbFitCont2, p.value = cutoff, method = "global")
summary(wtResCont2)
```

probes that have some expression changes up to P6:
```{r}
changed <- which(rowSums(abs(wtResCont2)) > 0)

(hits21 <- rownames(prDat)[changed][1:4])
makeStripplot(prepareData(hits21))
```

probes that hold steady between P5 all the way to 4_weeks:
```{r}
wtResCont3 <- decideTests(wtEbFitCont2, p.value = 1e-2, method = "global")
summary(wtResCont3)
steadyLater <- which(rowSums(abs(wtResCont3)) == 0)
```

```{r}
intersect(changed, steadyLater)
```

